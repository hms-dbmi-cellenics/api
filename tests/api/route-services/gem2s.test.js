/* eslint-disable arrow-body-style */
/* eslint-disable global-require */
const ExperimentService = require('../../../src/api/route-services/experiment');
const Gem2sService = require('../../../src/api/route-services/gem2s');
const SamplesService = require('../../../src/api/route-services/samples');
const ProjectsService = require('../../../src/api/route-services/projects');
const { createGem2SPipeline } = require('../../../src/api/general-services/pipeline-manage');

jest.mock('../../../src/api/route-services/experiment');
jest.mock('../../../src/api/route-services/samples');
jest.mock('../../../src/api/route-services/projects');
jest.mock('../../../src/api/general-services/pipeline-status');
jest.mock('../../../src/api/general-services/pipeline-manage');

const mockGem2sParamsBackendCall = (
  customProjectResponse = {},
  customSamplesResponse = {},
  customMetadataResponse = {},
) => {
  const projectResponse = {
    projectId: 'project-2',
    experimentName: 'New experiment',
    meta: {
      organism: null,
      type: '10x',
    },
    sampleIds: ['sample-1', 'sample-2'],
    ...customProjectResponse,
  };

  const samplesResponse = {
    samples: {
      'sample-1': { name: 'Sample 1' },
      'sample-2': { name: 'Sample 2' },
    },
    ...customSamplesResponse,
  };

  const metadataResponse = {
    metadataKeys: [],
    ...customMetadataResponse,
  };

  const mockSaveGem2sHandle = jest.fn();
  ExperimentService.mockImplementation(() => ({
    getExperimentData: jest.fn()
      .mockImplementationOnce(() => Promise.resolve(projectResponse)),
    saveGem2sHandle: mockSaveGem2sHandle,
  }));


  SamplesService.mockImplementation(() => ({
    getSamplesByExperimentId: jest.fn()
      .mockImplementationOnce(() => Promise.resolve(samplesResponse)),
  }));

  ProjectsService.mockImplementation(() => ({
    getProject: jest.fn()
      .mockImplementationOnce(() => Promise.resolve(metadataResponse)),
  }));

  return {
    mockSaveGem2sHandle,
  };
};

const experimentId = 'abcd1234';
const mockAuthJwt = 'mockAuthJwtToken';

describe('gem2s', () => {
  beforeEach(async () => {
    jest.resetModules();
    jest.clearAllMocks();
  });

  it('gem2sCreate - Creates new GEM2S handles and save them to database', async () => {
    const { mockSaveGem2sHandle } = mockGem2sParamsBackendCall();

    const mockBody = {
      paramsHash: 'gem2s-params-hash',
    };

    await Gem2sService.gem2sCreate(experimentId, mockBody, mockAuthJwt);

    // Create new handles
    expect(createGem2SPipeline).toHaveBeenCalled();

    // // Create new handles
    expect(mockSaveGem2sHandle).toHaveBeenCalled();
  });

  it('generateGem2sParams - Should generate GEM2S taskParams correctly', async () => {
    mockGem2sParamsBackendCall();

    const taskParams = await Gem2sService.generateGem2sParams(experimentId, mockAuthJwt);

    expect(taskParams).toMatchSnapshot();
  });

  it('sendUpdateToSubscribed - Should send update if payloads are correct', async () => {
    const mockedSocketsEmit = jest.fn();
    const mockIo = {
      sockets: {
        emit: mockedSocketsEmit,
      },
    };

    const parsedMessage = {
      taskName: 'downloadGem',
      experimentId: 'experimentId',
      authJWT: 'Bearer mockAuthJwtToken',
    };

    await Gem2sService.sendUpdateToSubscribed(experimentId, parsedMessage, mockIo);

    const emitParamsChannel = mockedSocketsEmit.mock.calls[0][0];
    expect(mockedSocketsEmit).toHaveBeenCalled();

    // Emitted to the correct channel
    expect(emitParamsChannel).toMatch(experimentId);
    expect(mockedSocketsEmit).toMatchSnapshot();
  });

  it('gem2sResponse - Should return message if message is valid', async () => {
    const mockedSocketsEmit = jest.fn();
    const mockIo = {
      sockets: {
        emit: mockedSocketsEmit,
      },
    };

    const validMessage = {
      taskName: 'downloadGem',
      experimentId,
    };

    await Gem2sService.gem2sResponse(mockIo, validMessage);

    const emitParamsChannel = mockedSocketsEmit.mock.calls[0][0];
    expect(mockedSocketsEmit).toHaveBeenCalled();

    // Emitted to the correct channel
    expect(emitParamsChannel).toMatch(experimentId);
    expect(mockedSocketsEmit).toMatchSnapshot();
  });


  it('gem2sResponse - Should return message if message is valid', async () => {
    const mockedSocketsEmit = jest.fn();
    const mockIo = {
      sockets: {
        emit: mockedSocketsEmit,
      },
    };

    const validMessage = {
      taskName: 'downloadGem',
      experimentId,
    };

    await Gem2sService.gem2sResponse(mockIo, validMessage);

    const emitParamsChannel = mockedSocketsEmit.mock.calls[0][0];
    expect(mockedSocketsEmit).toHaveBeenCalled();

    // Emitted to the correct channel
    expect(emitParamsChannel).toMatch(experimentId);
    expect(mockedSocketsEmit).toMatchSnapshot();
  });


  it('gem2sResponse - Should throw an error if message is invalid', async () => {
    const mockedSocketsEmit = jest.fn();
    const mockIo = {
      sockets: {
        emit: mockedSocketsEmit,
      },
    };

    const InvalidMessage = {
      taskName: 'downloadGem',
    };

    await expect(Gem2sService.gem2sResponse(mockIo, InvalidMessage)).rejects.toBeInstanceOf(Error);

    expect(mockedSocketsEmit).not.toHaveBeenCalled();
  });

  it('gem2sReponse - Pass on error message properly', async () => {
    const mockedSocketsEmit = jest.fn();
    const mockIo = {
      sockets: {
        emit: mockedSocketsEmit,
      },
    };

    const errorText = 'some unknwon error';

    const errorMessage = {
      experimentId,
      response: {
        error: errorText,
      },
    };

    await Gem2sService.gem2sResponse(mockIo, errorMessage);

    const emitParamsChannel = mockedSocketsEmit.mock.calls[0][0];
    const errorMessge = mockedSocketsEmit.mock.calls[0][1].response.error;

    expect(mockedSocketsEmit).toHaveBeenCalled();

    // Emitted to the correct channel
    expect(emitParamsChannel).toMatch(experimentId);
    expect(errorMessge).toEqual(errorText);

    expect(mockedSocketsEmit).toMatchSnapshot();
  });
});
