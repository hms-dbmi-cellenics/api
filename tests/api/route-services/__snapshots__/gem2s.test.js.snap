// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gem2s gem2sReponse - Pass on error message properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "ExperimentUpdates-abcd1234",
      Object {
        "experimentId": "abcd1234",
        "response": Object {
          "error": "some unknwon error",
        },
        "status": Object {
          "gem2s": Object {
            "completedSteps": Array [],
            "startDate": null,
            "status": "SUCCEEDED",
            "stopDate": null,
          },
        },
        "type": "gem2s",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gem2s gem2sResponse - Should return message if message is valid 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "ExperimentUpdates-abcd1234",
      Object {
        "experimentId": "abcd1234",
        "status": Object {
          "gem2s": Object {
            "completedSteps": Array [],
            "startDate": null,
            "status": "SUCCEEDED",
            "stopDate": null,
          },
        },
        "taskName": "downloadGem",
        "type": "gem2s",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gem2s gem2sResponse - Should return message if message is valid 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "ExperimentUpdates-abcd1234",
      Object {
        "experimentId": "abcd1234",
        "status": Object {
          "gem2s": Object {
            "completedSteps": Array [],
            "startDate": null,
            "status": "SUCCEEDED",
            "stopDate": null,
          },
        },
        "taskName": "downloadGem",
        "type": "gem2s",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`gem2s generateGem2sParams - Should generate GEM2S taskParams correctly 1`] = `
Object {
  "authJWT": "mockAuthJwtToken",
  "experimentName": "New experiment",
  "input": Object {
    "type": "10x",
  },
  "organism": null,
  "projectId": "project-2",
  "sampleIds": Array [
    "sample-1",
    "sample-2",
  ],
  "sampleNames": Array [
    "Sample 1",
    "Sample 2",
  ],
}
`;

exports[`gem2s sendUpdateToSubscribed - Should send update if payloads are correct 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "ExperimentUpdates-abcd1234",
      Object {
        "authJWT": "Bearer mockAuthJwtToken",
        "experimentId": "experimentId",
        "status": Object {
          "gem2s": Object {
            "completedSteps": Array [],
            "startDate": null,
            "status": "SUCCEEDED",
            "stopDate": null,
          },
        },
        "taskName": "downloadGem",
        "type": "gem2s",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
